drivers/serial/uart_ene_kb1200.c

```diff
@@ -0,0 +1,378 @@
+	.irq_tx_ready = kb1200_uart_irq_tx_ready,
+	.irq_rx_enable = kb1200_uart_irq_rx_enable,
+	.irq_rx_disable = kb1200_uart_irq_rx_disable,
+	.irq_rx_ready = kb1200_uart_irq_rx_ready,
+	.irq_err_enable = kb1200_uart_irq_err_enable,
+	.irq_err_disable = kb1200_uart_irq_err_disable,
+	.irq_is_pending = kb1200_uart_irq_is_pending,
+	.irq_update = kb1200_uart_irq_update,
+	.irq_callback_set = kb1200_uart_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+/* GPIO module instances */
+#define KB1200_UART_DEV(inst) DEVICE_DT_INST_GET(inst),
+static const struct device *const uart_devices[] = {DT_INST_FOREACH_STATUS_OKAY(KB1200_UART_DEV)};
+static void kb1200_uart_isr_wrap(const struct device *dev)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(uart_devices); i++) {
+		const struct device *dev_ = uart_devices[i];
+		const struct kb1200_uart_config *config = dev_->config;
+
+		if (config->ser->SERIE & config->ser->SERPF) {
+			kb1200_uart_irq_handler(dev_);
+		}
+	}
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static int kb1200_uart_init(const struct device *dev)
+{
+	int ret;
+	const struct kb1200_uart_config *config = dev->config;
+	struct kb1200_uart_data *data = dev->data;
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret != 0) {
+		return ret;
+	}
+
+	kb1200_uart_configure(dev, &data->current_config);
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	config->irq_cfg_func();
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+	return 0;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static bool init_irq = true;
+static void kb1200_uart_irq_init(void)
+{
+	if (init_irq) {
+		init_irq = false;
+		IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), kb1200_uart_isr_wrap, NULL,
+			    0);
+		irq_enable(DT_INST_IRQN(0));
+	}
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+#define KB1200_UART_INIT(n)                                                                        \
+	PINCTRL_DT_INST_DEFINE(n);                                                                 \
+	static struct kb1200_uart_data kb1200_uart_data_##n = {                                    \
+		.current_config = {                                                                \
+				.baudrate = DT_INST_PROP(n, current_speed),                        \
+				.parity = UART_CFG_PARITY_NONE,                                    \
+				.stop_bits = UART_CFG_STOP_BITS_1,                                 \
+				.data_bits = UART_CFG_DATA_BITS_8,                                 \
+				.flow_ctrl = UART_CFG_FLOW_CTRL_NONE,                              \
+			},                                                                         \
+	};                                                                                         \
+	static const struct kb1200_uart_config kb1200_uart_config_##n = {                          \
+		IF_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN, (.irq_cfg_func = kb1200_uart_irq_init,))  \
```

Comment:
this is not multi-instance capable, you're always using the same IRQ (from instance 0)

Reply:
The UARTs share the same IRQ and I have to distinguish them by the pending flags.

Reply:
ok, thanks for the explanation!